src/Psl/Xml/Issue/Level.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Xml\Issue;
+
+use Psl\Exception\InvariantViolationException;
+
+use function in_array;
+use function Psl\invariant;
+
+use const LIBXML_ERR_ERROR;
+use const LIBXML_ERR_FATAL;
+use const LIBXML_ERR_NONE;
+use const LIBXML_ERR_WARNING;
+
+/**
+ * @psalm-immutable
+ */
+final class Level
+{
+    private const LEVEL_ERROR = LIBXML_ERR_ERROR;
+    private const LEVEL_FATAL = LIBXML_ERR_FATAL;
+    private const LEVEL_NONE = LIBXML_ERR_NONE;
+    private const LEVEL_WARNING = LIBXML_ERR_WARNING;
+
+    private int $value;
+
+    private function __construct(int $value)
+    {
+        $this->value = $value;
+    }
+
+    /**
+     * @psalm-pure
+     * @throws InvariantViolationException
+     */
+    public static function fromLevelCode(int $value): self
+    {
+        invariant(
+            in_array($value, [self::LEVEL_ERROR, self::LEVEL_FATAL, self::LEVEL_NONE, self::LEVEL_WARNING], true),
+            'Received invalid XML issue level: ' . (string) $value
+        );
+
+        return new self($value);
+    }
+
+    /**
+     * @psalm-pure
+     */
+    public static function error(): self
+    {
+        return new self(self::LEVEL_ERROR);
+    }
+
+    /**
+     * @psalm-pure
+     */
+    public static function fatal(): self
+    {
+        return new self(self::LEVEL_FATAL);
+    }
+
+    /**
+     * @psalm-pure
+     */
+    public static function warning(): self
+    {
+        return new self(self::LEVEL_WARNING);
+    }
+
+    /**
+     * @psalm-pure
+     */
+    public static function none(): self
+    {
+        return new self(self::LEVEL_NONE);
+    }
```

Comment:
should not be considered an "issue"

Reply:
Decided to keep none since the collection can return Level::none() as highest level if it is empty. Makes sense from a level value object POV right?

Reply:
why not null? indicating there are no issues to pick the highest from :)

Reply:
Isn't it strange to get `null` back from `$issues->getHighestLevel()`?

Reply:
throw instead? we have `count()` so users should probably call that before starting to check the errors and the highest level anyway.

Reply:
you would most likely use it as

```php
if ($issues->getHighestLevel() > Level::warning()->value()) // ...
```

Would you prefer to

```php
try {
   if ($issues->getHighestLevel() > Level::warning()->value()) // ...
} catch () {
// ...
}
```

For no real reason? 


(Agreeing that you could use count() first ... but to make it dummy proof - returning 0 seems ok-ish ..)

Reply:
Like: :) 
```php
if (0 === $issues->count()) {
  return;
}

if ($issues->getHighestLevel() > Level::warning()->value()) {

}
```

another thing we can do is ... not return `IssuesCollecttion` unless we do actually have issues, otherwise null? that seems sensable, so that `IssueCollection` is always on a valid state that contains issues.

( throws in the constructor if no issues are passed ... )

Reply:
but that might bring an issue with the filter method throwing ... 

also to note, that return null works with `if ($issues->getHighestLevel() > Level::warning()->value())` since null will be cast to 0 :p

Reply:
In that case you should do

```php
if ($issues->getHighestLevel()?->value() > Level::warning()->value())
```

Since the issue is more something that is used internally, I really don't think adding an empty state is a bad idea.
Just for convenience. The ideas above all have there downsides:

- If you check for count 0, psalm still thinks it can throw
- if you only allow non empty, you have an issue filterting

Reply:
> to be compatible with libxm, which has a none error level - otherwise the internal level detector might fail as well.

we don't want compatible with libxml, or any other PHP Core extension, as most of them have broken APIs already. that's the point of this library :)

> If you check for count 0, psalm still thinks it can throw

`/** @psalm-suppress MissingThrowsDocblock - collection is not empty*/` does the job in this case, we already do this tons of times within PSL, because it's valid.